### PR TITLE
Parsing headers (including HeadersObject)

### PR DESCRIPTION
**Pull request by klokane**
Also contain generalised parseMap

### CODE REVIEW

File: packages/fury-adapter-oas3-parser/test/integration/fixtures/petstore.json [hunk 1]
```diff
@@ -107,11 +107,10 @@
                               "content": {
                                 "key": {
                                   "element": "string",
-                                  "content": "Content-Type"
+                                  "content": "x-next"
```
Comment (kylef) [inline packages/fury-adapter-oas3-parser/test/integration/fixtures/petstore.json]: This seems to be disclosing a bug where the headers object is overriding the `Content-Type`. I think the behaviour should be that the `Content-Type` should still be present but we should only add it if there isn't a user defined Content-Type header.

File: packages/fury-adapter-oas3-parser/lib/parser/oas/parseHeaderObject.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+const pipeParseResult = require('../../pipeParseResult');
+const parseObject = require('../parseObject');
+
+const name = 'Header Object';
+const unsupportedKeys = [
+  'description', 'required', 'deprecated', 'allowEmptyValue',
+];
+const isUnsupportedKey = R.anyPass(R.map(hasKey, unsupportedKeys));
+
+
```
Comment (pksunkara) [inline packages/fury-adapter-oas3-parser/lib/parser/oas/parseHeaderObject.js]:
```suggestion
```

File: packages/fury-adapter-oas3-parser/lib/parser/parseMap.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
+const R = require('ramda');
+const pipeParseResult = require('../pipeParseResult');
+const {
+  isObject, getValue, isAnnotation,
+} = require('../predicates');
+const { createWarning } = require('./annotations');
+const parseObject = require('./parseObject');
+
+const validateMapValueIsObject = (namespace, name, key) => R.unless(R.pipe(getValue, isObject), createWarning(namespace, `'${name}' value of key '${key}' is not an object`));
```
Comment (pksunkara) [inline packages/fury-adapter-oas3-parser/lib/parser/parseMap.js]:
```suggestion
const validateMapValueIsObject = (namespace, name, key) => R.unless(R.pipe(getValue, isObject), createWarning(namespace, `'${name}' '${key}' is not an object`));
```

File: packages/fury-adapter-oas3-parser/test/unit/parser/oas/parseComponentsObject-test.js [hunk 4]
```diff
@@ -202,25 +202,82 @@ describe('Components Object', () => {
...
-    it('provides warning for unsupported headers key', () => {
+    it('it provide warning, if headers is not map', () => {
       const components = new namespace.elements.Object({
-        headers: {},
+        headers: 'dummy',
+      });
+
+      const parseResult = parse(context, components);
+
+      expect(parseResult).to.contain.warning("'Components Object' value of key 'headers' is not an object");
```
Comment (pksunkara) [inline packages/fury-adapter-oas3-parser/test/unit/parser/oas/parseComponentsObject-test.js]:
```suggestion
      expect(parseResult).to.contain.warning("'Components Object' 'headers' is not an object");
```

File: packages/fury-adapter-oas3-parser/test/unit/parser/oas/parseResponseObject-test.js [hunk 5]
```diff
@@ -64,18 +64,99 @@ describe('Response Object', () => {
...
+    it('provides warning, if headers is not object', () => {
       const response = new namespace.elements.Object({
         description: 'response 200',
         headers: 'dummy',
       });
 
       const parseResult = parse(context, response);
 
-      expect(parseResult).to.contain.warning("'Response Object' contains unsupported key 'headers'");
+      expect(parseResult).to.contain.warning("'Response Object' value of key 'headers' is not an object");
```
Comment (pksunkara) [inline packages/fury-adapter-oas3-parser/test/unit/parser/oas/parseResponseObject-test.js]:
```suggestion
      expect(parseResult).to.contain.warning("'Response Object' 'headers' is not an object");
```

File: packages/fury-adapter-oas3-parser/test/unit/parser/oas/parseResponseObject-test.js [hunk 6]
```diff
@@ -64,18 +64,99 @@ describe('Response Object', () => {
     expect(parseResult).to.contain.warning("'Response Object' is not an object");
   });
 
-  describe('warnings for unsupported properties', () => {
-    it('provides warning for unsupported headers key', () => {
+  describe('#headers', () => {
+    it('provides warning, if headers is not object', () => {
```
Comment (pksunkara) [inline packages/fury-adapter-oas3-parser/test/unit/parser/oas/parseResponseObject-test.js]:
```suggestion
    it('provides warning when headers is not an object', () => {
```

File: packages/fury-adapter-oas3-parser/test/unit/parser/parseMap-test.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  beforeEach(() => {
+    context = new Context(namespace);
+  });
+
+  it('will report if map is not object', () => {
+    const member = new context.namespace.elements.Member('key', 'value');
+
+    const result = parseMap(context, 'dummy', 'key', parseDummy)(member);
+
+    expect(result).to.contain.warning("'dummy' value of key 'key' is not an object");
```
Comment (pksunkara) [inline packages/fury-adapter-oas3-parser/test/unit/parser/parseMap-test.js]:
```suggestion
    expect(result).to.contain.warning("'dummy' 'key' is not an object");
```

File: packages/fury-adapter-oas3-parser/test/integration/fixtures/petstore.json (hunk 1 above)
Comment (klokane) [reply packages/fury-adapter-oas3-parser/test/integration/fixtures/petstore.json]: Fixed

File: packages/fury-adapter-oas3-parser/test/unit/parser/oas/parseComponentsObject-test.js [hunk 8]
```diff
@@ -202,25 +202,82 @@ describe('Components Object', () => {
...
-        headers: {},
+        headers: 'dummy',
+      });
+
+      const parseResult = parse(context, components);
+
+      expect(parseResult).to.contain.warning("'Components Object' 'headers' is not an object");
+    });
+
+    it('provide warning if header is not HeaderObject', () => {
```
Comment (pksunkara) [inline packages/fury-adapter-oas3-parser/test/unit/parser/oas/parseComponentsObject-test.js]:
```suggestion
    it('provide warning when header is non-object', () => {
```

File: packages/fury-adapter-oas3-parser/test/unit/parser/oas/parseComponentsObject-test.js [hunk 9]
```diff
@@ -202,25 +202,82 @@ describe('Components Object', () => {
...
       });
 
       const parseResult = parse(context, components);
 
-      expect(parseResult).to.contain.warning("'Components Object' contains unsupported key 'examples'");
+      expect(parseResult).to.not.contain.annotations;
     });
 
-    it('provides warning for unsupported headers key', () => {
+    it('it provide warning, if headers is not map', () => {
```
Comment (pksunkara) [inline packages/fury-adapter-oas3-parser/test/unit/parser/oas/parseComponentsObject-test.js]:
```suggestion
    it('provides warning when waders is non-object', () => {
```

**Pull request merged**